### PR TITLE
added AI

### PR DESCRIPTION
**Pull request by plasne**
# Pull Request Template

## What are you trying to address

- Added docs on AI tooling that is available and how and when our devcrew found it valuable.

### CODE REVIEW

**Review by kavithasreenath — APPROVED** [review]
LGTM

**Review by TessFerrandez — COMMENTED** [review]
There seems to be a fair bit of overlap between this content and the content in code completion.
Can we perhaps build on the content that is already there in code completion, and add what we see missing. I very much like to have the content under the label code completion vs AI as code completion shows why it is a part of the developer experience section.

On that topic, the section on prompt engineering here, feels a bit out of place in the section of developer experience, thoughts?

**Comment by plasne** [comment]
> Great content, thank you for the contribution...
> 
> There seems to be a fair bit of overlap between this content and the content in code completion. Can we perhaps build on the content that is already there in code completion, and add what we see missing. I very much like to have the content under the label code completion vs AI as code completion shows why it is a part of the developer experience section.
> 
> On that topic, the section on prompt engineering here, feels a bit out of place in the section of developer experience, thoughts?

I don't really agree. Code completion is one aspect of how you can use AI tooling and we go into a lot more than that in the AI article - bug fixes, documentation, explaining code, improving reliability, writing secondary components (like indexes), etc. are all other aspects. I left code completion alone even though there is some overlap because I think that is one common scenario, but far from the only scenario.

Prompt engineering was added by a colleague of mine, it was not originally part of the article, however, I don't know of a better place to put it. His point for having it here is that the quality of the AI output is directly impacted by the input and there is a lot that can be done on the prompt engineering side to improve the output. In that I agree, I have seen first hand that simply asking something a slightly different way produces significantly different output.

I am happy to discuss further, but I don't think reducing AI to code completion is the right way to go. Copilot v1 was a code completion tool, but subsequent versions and features added to those versions are mostly about other things.

**Comment by TessFerrandez** [comment]
Let's discuss offline. I am not disputing that there are other aspects of AI than code completion. I am however failing to see how they relate to developer experience or engineering fundamentals so would love to hear your thoughts on that. 

Regarding the duplicated content. If we want to add something about copilot, We definitely don't want to duplicate content as it gets confusing and we have to maintain in multiple places. We also want to be cautious of duplicating too much content from the product pages that is prone to change, like the release info. Better keep it short and sweet and link instead.

File: docs/developer-experience/artificial-intelligence.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+Copilot has extension that offers additional features that are not yet ready for prime-time. You can install it from the VS Code Marketplace. These features include:
+
+- __Explain__. Copilot can explain what the code is doing in natural language.
+- __Translate__. Copilot can translate code from one language to another.
+- __Brushes__. You can select code that Copilot then modifies inline based on a "brush" you select, for example, to make the code more readable, fix bugs, improve debugging, document, etc.
+- __Generate Tests__. Copilot can generate unit tests for your code. Though currently this is limited to JavaScript and TypeScript.
+
+## GitHub Copilot X
+
+This extention for Copilot is currently in technical preview. It will enable the following additional features:
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:28]:
```suggestion
This extension for Copilot is currently in technical preview. It will enable the following additional features:
```

File: docs/developer-experience/artificial-intelligence.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# Artificial Intelligence
+
+There are a number of AI tools that can improve the developer experience. This article will discuss tooling that is available as well as advice on when it might be appropriate to use such tooling.
+
+## GitHub Copilot
+
+The current version of GitHub Copilot can provide code completion in Visual Studio Code. It is a VS Code extension that can be installed from the VS Code Marketplace. It is currently in beta and requires a GitHub account to use. It is free to use during the beta period. It is currently available for VS Code only, but there are plans to support other IDEs in the future. It is currently available for the following languages: JavaScript, TypeScript, Ruby, Go, Python, C#, C++, and Java. You can find more information about GitHub Copilot [here](https://copilot.github.com/).
+
+Developers on our team have found the following use-cases for GitHub Copilot:
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:9]: Can we reword this (and similar statements below). This wording fits great in blogs but feels a bit out of place in the playbook. Perhaps something like... "some example use-cases are"

File: docs/developer-experience/artificial-intelligence.md (hunk 1 above)
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:28]: If we describe the availability, this page will immediately be outdated when the component is released. To avoid this, change this and say, see ... for availability

File: docs/developer-experience/artificial-intelligence.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+- __Write Documentation__. For example, the above paragraph was written using Copilot.
+- __Write Unit Tests__. Given that setup and assertions are often consistent across unit tests, Copilot tends to be very accurate.
+- __Unblock__. It is often hard start writing when staring at a blank page, Copilot can fill the space with something that may or may not be what you ultimately want to do, but it can help get you in the right headspace.
+
+If you want Copilot to write something useful for you, try writing a comment that describes what your code is going to do - it can often take it from there.
+
+## GitHub Copilot Labs
+
+Copilot has extension that offers additional features that are not yet ready for prime-time. You can install it from the VS Code Marketplace. These features include:
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:19]:
```suggestion
Copilot has an extension that offers additional features that are not yet ready for prime-time. You can install it from the VS Code Marketplace. These features include:
```

Would be good to point to the extension page

File: docs/developer-experience/artificial-intelligence.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+- __Write Code__. Given prompting by the developer it can write code that you can one-click deploy into existing or new files.
+- __Debug__. Copilot can analyze your code and propose solutions to fix bugs.
+
+It can do most of what Labs can do with "brushes" as "topics", but whereas Labs changes the code in your file, the chat functionality just shows what it would change in the window. However, there is also an "inline mode" for GitHub Copilot Chat that allows you to make changes to your code inline which does not have this same limitation.
+
+You must sign-up for the preview [here](https://github.com/features/preview/copilot-x). Once you are accepted, you will be able to install the extension from the VS Code Marketplace. You may require the Insiders build of VS Code.
+
+## ChatGPT / Bing Chat
+
+Prior to GitHub Copilot X chat availability, we used ChatGPT and Bing Chat to provide similar functionality. We found the generation of code using ChatGPT to be most valuable for these scenarios:
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:41]: Same comment as before about wording, change the we... to some scenarios are... or something to that effect

File: docs/developer-experience/artificial-intelligence.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+- __Build SQL Indexes__. Given a query, ChatGPT can generate a SQL index that will improve the performance of the query.
+- __Write Regular Expressions__. These are notoriously difficult to write, but ChatGPT can generate them for you if you give some sample input and describe what you want to extract.
+- __Improve and Validate__. If you are unsure of the implications of writing code a particular way, you can ask ChatGPT questions about it. For instance, you might ask if there is a way to write the code that is more performant or uses less memory. Once it gives you an opinion, you can ask it to provide documentation validating that assertion.
+- __Change Perspective__. ChatGPT can impersonate a persona or even a system and answer questions from that perspective. For example, you can ask it to explain what a particular piece of code does from the perspective of a user. You might have ChatGPT imagine it is a database adminstrator and ask it to explain how to improve a particular query.
+
+When using Bing Chat, we found much better results using Creative Mode.
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:49]:
```suggestion
When using Bing Chat, experiment with modes, sometimes changing to creative mode can give the results you need.
```

**Review by TessFerrandez — CHANGES_REQUESTED** [review]
I had a second look at this, and I see now how this might merit its own page... I think however that we should find a better name than the very generic AI as this will greatly reduce the discoverability of the content. A page on AI under dev experience can be about anything, including dev experience for ai systems. 

Perhaps a title like copilots can r similar would be better. 

Also, in order to avoid document duplication. Please remove the code completion page as part of this PR. If we have a full page on copilots, the code completion page is no longer needed.

File: docs/developer-experience/artificial-intelligence.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+## Prompt Engineering
+
+Prompt Engineering is a cutting edge subject area focused on prompting strategies designed to derive the most value from an AI response.
+
+Most LLM services such as GPT-4 or Bing Chat have a limited number of prompts a user can send to it in a given amount of time. Even outside of that scenario, we want to get the most useful information as quickly as possible.
+
+This is one open source example of documentation around prompt engineering, and should be seen as a jumping off point for further investigation. [Link](https://github.com/brexhq/prompt-engineering)
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:57]:
```suggestion
To get the best results with the copilot tools, a good prompt is crucial.

This repo can help serve as a starting point for learning and experimenting with prompt engineering. [Link](https://github.com/brexhq/prompt-engineering)
```

On first read, it was hard to follow how this related to the rest of the text or to dev experience. 

Suggestion: add context - see example above, and perhaps remove some of of the details about prompt engineering, that while true, are not necessary for the given content. We want to keep the playbook succinct and to the point, so that it is easy to find the content you need.

File: docs/developer-experience/artificial-intelligence.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+## Considerations
+
+It is important when using AI tools to understand how the data (including private or commercial code) might be used by the system. GitHub Copilot provides guarantees on how data will be used that should align with customer expectations, however, we should always seek a customer's consent. More information on the privacy can be found [here](https://resources.github.com/copilot-for-business/).
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:61]: Again here, link to the page, and let them do the guarantees. If we make statements around guarantees on this page, we bind ourselves to update this if it ever changes. Else we will have two conflicting Microsoft sources on the topic.

File: docs/developer-experience/artificial-intelligence.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+ChatGPT allows data entered into the prompts to be used for training, so private or commercial code should not be copied into that system.
```
Comment (TessFerrandez) [inline docs/developer-experience/artificial-intelligence.md:63]: Same as above...

**Review by TessFerrandez — CHANGES_REQUESTED** [review]

File: docs/developer-experience/artificial-intelligence.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
+# Artificial Intelligence
```
Comment (shiranr) [inline docs/developer-experience/artificial-intelligence.md:1]: I feel the title is misleading, I think it needs to be Copilots and not AI as AI is a big topic 
```suggestion
# Copilots
```

File: docs/developer-experience/artificial-intelligence.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+- __Chat__. Rather than just providing code completion, Copilot will be able to have a conversation with you about what you want to do. It has context about the code you are working on and can provide suggestions based on that context.
+- __Explain__. Copilot can explain what the code is doing in natural language.
+- __Write Code__. Given prompting by the developer it can write code that you can one-click deploy into existing or new files.
+- __Debug__. Copilot can analyze your code and propose solutions to fix bugs.
+
+It can do most of what Labs can do with "brushes" as "topics", but whereas Labs changes the code in your file, the chat functionality just shows what it would change in the window. However, there is also an "inline mode" for GitHub Copilot Chat that allows you to make changes to your code inline which does not have this same limitation.
+
+You must sign-up for the preview [here](https://github.com/features/preview/copilot-x). Once you are accepted, you will be able to install the extension from the VS Code Marketplace. You may require the Insiders build of VS Code.
+
+## ChatGPT / Bing Chat
```
Comment (shiranr) [inline docs/developer-experience/artificial-intelligence.md:39]: I would consider focusing this page on copilots and create another content for ChatGPT/Bing Chat.

Comment (TessFerrandez) [reply docs/developer-experience/artificial-intelligence.md:39]: I agree on focusing the page on copilots... 
Regarding placing ChatGPT/Bing Chat elsewhere, @shiranr do you feel that content fits in the EF playbook, or did you mean perhaps elsewhere as in not in the playbook?

Comment (shiranr) [reply docs/developer-experience/artificial-intelligence.md:39]: This specific content is not relevant to the EF playbook but if people are using ChatGPT / Bing Chat for programming and can give a solid to do/ not to do about it, it MIGHT be good content for the playbook

File: docs/developer-experience/artificial-intelligence.md (hunk 2 above)
Comment (shiranr) [reply docs/developer-experience/artificial-intelligence.md:9]: Agreed, the playbook should be more about hard cold facts.

Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md:9]: Yes, we can. However, realize that this content comes from the "sharing sprint" for a devcrew. We are not the product group saying facts about what use-cases are supported by GitHub Copilot. We are a devcrew who used the tools in an engagement - it can do a lot of things - these are the things that we as a crew found productive. Our guiding principal for sharing, in fact, has been that we don't share anything that we haven't actually vetted and proven to be useful on our team. If we just say "some example use-cases" then it seems like we are losing this important distinction.

File: docs/developer-experience/artificial-intelligence.md (hunk 1 above)
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md:28]: I think we may just have to update it once it becomes available. How and where to sign-up and how to check whether you are "in" was really hard for folks. I had several people reach out to me for help, so I put the links here.

File: docs/developer-experience/artificial-intelligence.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md:41]: For one, not everyone in our crew has been granted access to Copilot X yet. Secondly, Copilot will only answer "questions about programming", sometimes legitimate questions seem to be caught by that filter and we still use ChatGPT or Bing Chat for that.

I will add this context.

File: docs/developer-experience/artificial-intelligence.md (hunk 6 above)
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md:57]: I made it less verbose.

File: docs/developer-experience/artificial-intelligence.md (hunk 2 above)
Comment (TessFerrandez) [reply docs/developer-experience/artificial-intelligence.md:9]: By sharing it in the EF playbook, the connotation is already that someone in CSE has found it useful, no need to call it out here.

File: docs/developer-experience/artificial-intelligence.md (hunk 7 above)
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md:61]: This doesn't say what those guarantees are.

One problem we had with ChatGPT is we couldn't provide it any prompts that included sensitive code. However, Copilot is designed to handle that scenario. Can you provide another way to say that?

File: docs/developer-experience/artificial-intelligence.md (hunk 1 above)
Comment (TessFerrandez) [reply docs/developer-experience/artificial-intelligence.md:28]: If there is an issue where people can't tell from the product page how to sign up, we should make an effort to connect with the product group to make this easier.

File: docs/developer-experience/artificial-intelligence.md (hunk 5 above)
Comment (TessFerrandez) [reply docs/developer-experience/artificial-intelligence.md]: We generally try to avoid blog/story telling wording like "our dev crew bla bla..." to keep a consistent tone of voice, also note that the playbook, while written by dev crews is meant both for our crews, customers and partners, so we should keep our internal structures out of the playbook and use more imperative language.

File: docs/developer-experience/artificial-intelligence.md (hunk 3 above)
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md]: done.

File: docs/developer-experience/artificial-intelligence.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TessFerrandez) [reply docs/developer-experience/artificial-intelligence.md]: As long as we keep it consistent with the topic so that it makes sense in the context of dev experience that is cool. We don't want it to scope creep and turn into a page about all things AI at Microsoft.

File: docs/developer-experience/artificial-intelligence.md (hunk 1 above)
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md]: As discussed, removing this section. I will consider if this is better for the wiki or just point people to the official documentation.

File: docs/developer-experience/artificial-intelligence.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+- __Chat__. Rather than just providing code completion, Copilot will be able to have a conversation with you about what you want to do. It has context about the code you are working on and can provide suggestions based on that context.
+- __Explain__. Copilot can explain what the code is doing in natural language.
+- __Write Code__. Given prompting by the developer it can write code that you can one-click deploy into existing or new files.
+- __Debug__. Copilot can analyze your code and propose solutions to fix bugs.
+
+It can do most of what Labs can do with "brushes" as "topics", but whereas Labs changes the code in your file, the chat functionality just shows what it would change in the window. However, there is also an "inline mode" for GitHub Copilot Chat that allows you to make changes to your code inline which does not have this same limitation.
+
+You must sign-up for the preview [here](https://github.com/features/preview/copilot-x). Once you are accepted, you will be able to install the extension from the VS Code Marketplace. You may require the Insiders build of VS Code.
```
[the remark this answers is not in the document]
Comment (plasne) [reply docs/developer-experience/artificial-intelligence.md]: removed.

**Review by TessFerrandez — APPROVED** [review]
Looks great

**Review by shiranr — APPROVED** [review]
